File: tesla/hash table/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    if(s.length=== 0 || s===null) return 0;
    let set = new Set<string>();
    let max_length: number = 0; // FIXED: Proper TS type annotation
    let left: number = 0;
    let n: number = s.length;
    for(let i=0; i<n; i++){
        while(set.has(s.charAt(i))){
            set.delete(s.charAt(left))
            
            left++
        }

        set.add(s.charAt(i));
        max_length= Math.max(max_length,set.size)
    }

    return max_length;
}

File: tesla/hash table/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length()==0 || s==null) return 0 ;
        int n =s.length();
        Set<Character> set = new HashSet<>();
        int max_length = Integer.MIN_VALUE;
        int left = 0;
        for(int i =0; i<n; i++){
            
            while(set.contains(s.charAt(i))){
                set.remove(s.charAt(left));
                left++;
            }

            set.add(s.charAt(i));
            max_length= Math.max(max_length, set.size());
        }

        return max_length;
    }
}